### ISSUE TITLE
Bug: Fetch Adapter support for Request objects is incomplete

### ISSUE THREAD

**Issue by fudomunro** [Netflix/pollyjs#255]
As of https://github.com/Netflix/pollyjs/pull/220, the Fetch Adapter allows you pass in a Request object as the only argument. However, it will simply extract the URL from the Request, and not use any of the other parameters.

This means if I have code being tested that uses Request objects to POST with a bunch of headers, like this:
```
const reqOpts = {
    body: xml,
    method: "POST",
    headers: {
      Accept: "application/xml",
      "Content-Type": "application/xml",
      "X-custom-header": "secret value"
    }
  };
req = new Request(url, reqOpts);
fetch(req);
```

Polly will do a GET to `url`, with a minimal set of headers.

I suspect this could be mostly fixed by modifying this code:
https://github.com/Netflix/pollyjs/blob/master/packages/%40pollyjs/adapter-fetch/src/index.js#L41

To be:
```
      // Support Request object
      if (typeof url === 'object' && 'url' in url) {
        options = url;
        url = url.url; 
      }
```

I'm testing this solution for us, I'll let you know how it goes.

**Comment by fudomunro** [comment]
This doesn't seem to be enough. It's maintaining the method now, but not the headers. I'll keep working on that.

**Comment by fudomunro** [comment]
I figured out what was going on with the headers, and got that sorted out, but there's still something very fishy going on.

If I set Polly to be in passthrough mode, and change our app code to call `fetch(url, options)` instead of `fetch(Request(url, options))`, it works. But using the developer tools, I grabbed the HAR of this fetch for each run, and they are identical (other than the responses).

**Comment by jasonmit** [comment]
Yeah looks like we're not doing enough to support the Request object in it's entirety.

**Comment by fudomunro** [comment]
I believe in this case we are using a session cookie for authentication, and I see we call `fetch()` with `credentials: "include"`, which I believe is how that is supposed to be passed through. Perhaps that's where this is breaking down currently.

**Comment by fudomunro** [comment]
OK, I've sort of found where this is breaking down. Sort of, because only in the compiled version, where it uses `objectSpread()` like this:
```javascript
return this.native.apply(global$1, [pollyRequest.url, objectSpread({}, options, {
    method: pollyRequest.method,
    headers: pollyRequest.headers,
    body: pollyRequest.body
})]);
```

When `options` is a Request object, this doesn't work, it loses things like `credentials: "include"`. I'll try serializing the Request object before passing it in as options.

**Comment by fudomunro** [comment]
Turning the Request object into an anonymous object mostly worked, but it doesn't account for the body, which isn't a property, but something that must be read asynchronously. Still trying to figure out the best way to do that here.

**Comment by fudomunro** [comment]
OK, so I've rewritten the fetch method to turn everything into a `Request`, and do the request handling in the body of the returned promise. This is within the compiled version, because that's the only one I'm able to modify and see my changes reflected, but I'll get someone here who does JS to help me put together a nicer version for the source.

I've tested this a bit, and it seems to be performing well, but I'm a little concerned about using `.text()` for reading the body of the request, given that it provides 4 other methods.
```javascript
context.fetch = function(request){
  // If two arguments were passed in, we create a request object, treating the first as the URL, second as options
  if (arguments.length > 1 && arguments[1] !== undefined){
    request = new Request(request, arguments[1]);
  }
  // If one argument is passed in, and it's a string, we treat it as the URL and create a request object
  else if (_typeof_1(request) === 'string'){
    request = new Request(request);
  }
  // Otherwise the single argument is already the request we are looking for

  // Extract all the request properties into an options object:
  var options = {}
  for (var property in request){
    options[property] = request[property];
  }

  // Turn the headers from an iterator into an object
  var headers = serializeHeaders(options.headers);
  
  var promise = new promise$1(function (resolve, reject) {
    var respond = function respond(_ref) {
      var response = _ref.response,
          error = _ref.error;
      if (error) {
        reject(error);
      } else {
        resolve(response);
      }
    };
    
    request.text().then(function(myText){
      _this.handleRequest({
        url: request.url,
        method: request.method,
        headers: serializeHeaders(new context.Headers(headers)),
        body: myText.length > 0 ? myText : undefined,
        requestArguments: {
          options: options,
          respond: respond
        }
      });
    });
  });
  return promise;
};
```

**Comment by offirgolan** [comment]
I’ve played around with the idea of defaulting to using the Request object in order to support this but ran into some regressions with the generated for the Polly request ID changing. I do think this would be the best way to do this though. 

@jasonmit is working on some regression tests in order to catch these sort of issues. I’ll be back from vacation next week and look into this more.

**Comment by jasonmit** [comment]
I'll be honest, not terribly familiar with the `Request` API but the support between the options might mean it's not the best internal default.

https://developer.mozilla.org/en-US/docs/Web/API/Request#Browser_compatibility

> @jasonmit is working on some regression tests in order to catch these sort of issues. I’ll be back from vacation next week and look into this more.

Good to know this is related, I'll try and land that soon.

**Comment by offirgolan** [comment]
Fix released with `v2.6.3`.

**Comment by fudomunro** [comment]
Thanks!

**Comment by fudomunro** [comment]
Looks like it didn't quite work, I'm getting this:
```
index.esm.js:38 Uncaught TypeError: Failed to construct 'Request': Please use the 'new' operator, this DOM object constructor cannot be called as a function.
    at new ExtendedRequest (pollyjs-adapter-fetch.js:6502)
    at index.esm.js:22
    at fetchJSON (fetch.js:9)
    at getStatus (Status.js:10)
    at fetchAndCache (useFetch.js:35)
    at useFetch.js:145
    at commitHookEffectList (react-dom.development.js:17079)
    at commitPassiveHookEffects (react-dom.development.js:17103)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at HTMLUnknownElement.b (index.esm.js:4)
```

The line in question in the compiled version is:
```
_this2 = possibleConstructorReturn(this, getPrototypeOf$2(ExtendedRequest).call(this, url, options));
```

Which I believe corresponds to https://github.com/Netflix/pollyjs/blob/master/packages/@pollyjs/adapter-fetch/src/index.js#L49 :
```
super(url, options);
```

**Comment by offirgolan** [comment]
Can you provide a reproduction? Also, are you using the fetch adapter in the browser or in node?

**Comment by fudomunro** [comment]
I'm doing it in the browser. I'll see if I can provide a simple reproduction.

**Comment by dheeraj-jn** [comment]
@offirgolan @fudomunro here is a simple reproduction of the issue:
https://codesandbox.io/s/late-surf-xil0t

**Comment by jasonmit** [comment]
Actually, lets open this up as a new bug since it's not the same as originally reported.

**Comment by jasonmit** [comment]
FYI @offirgolan fixed this in 3.0.1

**Comment by fudomunro** [comment]
Excellent, I'll have to try this out again when I get back to work. Thanks everyone!
